### PR TITLE
topology: sof-tgl-sdw-max98373-rt5682: add support for ADL

### PR DESCRIPTION
**Pull request by Vamshigopal**
For speakers SDW2 link is used.
For Headset SDW0 link is used.
add sof-adl-sdw-max98373-rt5682 topology

Signed-off-by: Vamshi Krishna <vamshi.krishna.gopal@intel.com>

### CODE REVIEW

File: tools/topology/sof-tgl-sdw-max98373-rt5682.m4 [hunk 1]
```diff
@@ -34,6 +34,13 @@ DEBUG_START
 #              |
 #              |
 # PCM3 <---- demux  <----- ALH0x103 (Speaker -max98373)
+ifelse(PLATFORM, `adl', `
+# PCM2 ----> smart_amp ----> ALH0x202 (Speaker -max98373)
+#               ^
+#               |
+#               |
+# PCM3 <---- demux  <----- ALH0x203 (Speaker -max98373)
+', `')
```
Comment (kv2019i) [inline tools/topology/sof-tgl-sdw-max98373-rt5682.m4:43]: I'd remove the ifelse section and just have something generic in the document like:
```
# PCM2 ----> smart_amp ----> ALH0xy02 (Speaker -max98373)
#              ^
#              |
#              |
# PCM3 <---- demux  <----- ALH0xy03 (Speaker -max98373)
```

**Review by kv2019i — CHANGES_REQUESTED** [review]
Thanks @Vamshigopal ! Now the documentation ends up duplicated, so I'd fix that before merge.

File: tools/topology/sof-tgl-sdw-max98373-rt5682.m4 (hunk 1 above)
Comment (Vamshigopal) [reply tools/topology/sof-tgl-sdw-max98373-rt5682.m4:43]: Ok sure, sound good! . Thanks @kv2019i

File: tools/topology/sof-tgl-sdw-max98373-rt5682.m4 [hunk 2]
```diff
@@ -51,6 +58,11 @@ define(`SMART_ALH_INDEX', 0x102)
 #define ALH BE dai_link name
 define(`SMART_ALH_PLAYBACK_NAME', `SDW1-Playback')
 define(`SMART_ALH_CAPTURE_NAME', `SDW1-Capture')
+ifelse(PLATFORM, `adl',
```
[the remark this answers is not in the document]
Comment (Vamshigopal) [reply tools/topology/sof-tgl-sdw-max98373-rt5682.m4:61]: If i add ifelse for  sdw1-playback case , that will break the current TGL topology.

[the remark this answers is not in the document]
Comment (plbossart) [reply tools/topology/sof-tgl-sdw-max98373-rt5682.m4:61]: what @RanderWang meant is that you cannot redefine a macro for ADL

````m4
define(`SMART_ALH_CAPTURE_NAME', `SDW1-Capture')
ifelse(PLATFORM, `adl',
	define(`SMART_ALH_CAPTURE_NAME', `SDW2-Capture')
````

that's not good practice, you need to have a clean if/else branch for TGL and ADL.

**Review by plbossart — CHANGES_REQUESTED** [review]

File: tools/topology/sof-tgl-sdw-max98373-rt5682.m4 (hunk 2 above)
[the remark this answers is not in the document]
Comment (Vamshigopal) [reply tools/topology/sof-tgl-sdw-max98373-rt5682.m4:61]: Ack, Thank you will update necessary changes

File: tools/topology/CMakeLists.txt [hunk 3]
```diff
@@ -134,8 +134,9 @@ set(TPLGS
 	"sof-tgl-max98373-rt5682\;sof-tgl-max98373-rt5682-xperi\;-DAMP_SSP=1\;-DINCLUDE_IIR_EQ=1"
 	"sof-tgl-max98373-rt5682\;sof-tgl-rt5682-ssp0-max98373-ssp2\;-DAMP_SSP=2"
 	"sof-tgl-max98373-rt5682\;sof-tgl-rt5682-ssp0-max98373-ssp2-xperi\;-DAMP_SSP=2\;-DINCLUDE_IIR_EQ=1"
-	"sof-tgl-sdw-max98373-rt5682\;sof-tgl-sdw-max98373-rt5682-2ch\;-DCHANNELS=2"
-	"sof-tgl-sdw-max98373-rt5682\;sof-tgl-sdw-max98373-rt5682-4ch\;-DCHANNELS=4"
+	"sof-tgl-sdw-max98373-rt5682\;sof-tgl-sdw-max98373-rt5682-2ch\;-DCHANNELS=2\;-DPLATFORM=tgl"
+	"sof-tgl-sdw-max98373-rt5682\;sof-tgl-sdw-max98373-rt5682-4ch\;-DCHANNELS=4\;-DPLATFORM=tgl"
+	"sof-tgl-sdw-max98373-rt5682\;sof-adl-sdw-max98373-rt5682\;-DCHANNELS=4\;-DPLATFORM=adl"
```
[the remark this answers is not in the document]
Comment (plbossart) [reply tools/topology/CMakeLists.txt:139]: not sure what the link was @sathya-nujella, I don't see a suggestion.

For other platforms we use link information but that's when we have multiple configurations per platforms, e.g. for TGL we must have 7 configurations already. For ADL you still have one, so no real pressure to change things.

File: tools/topology/sof-tgl-sdw-max98373-rt5682.m4 (hunk 2 above)
[the remark this answers is not in the document]
Comment (Vamshigopal) [reply tools/topology/sof-tgl-sdw-max98373-rt5682.m4:56]: I think there is a logic leak , can i use this way @plbossart @RanderWang ?
I will remove ifelse for tgl and add in else part of adl.

ifelse(PLATFORM, `adl',
` define(`SMART_ALH_INDEX', 0x202)
define(`SMART_ALH_PLAYBACK_NAME', `SDW2-Playback')
define(`SMART_ALH_CAPTURE_NAME', `SDW2-Capture')
',
` define(`SMART_ALH_INDEX', 0x102)
define(`SMART_ALH_PLAYBACK_NAME', `SDW1-Playback')
define(`SMART_ALH_CAPTURE_NAME', `SDW1-Capture')
')
This way there is no logic leak from original code which defined to use  SMART_ALH_PLAYBACK_NAME for any platform.

**Comment by lgirdwood** [comment]
@kv2019i @juimonen please review for BT.

File: tools/topology/sof-tgl-sdw-max98373-rt5682.m4 [hunk 4]
```diff
@@ -51,6 +58,11 @@ define(`SMART_ALH_INDEX', 0x102)
...
+`	define(`SMART_ALH_INDEX', 0x202)
+	define(`SMART_ALH_PLAYBACK_NAME', `SDW2-Playback')
+	define(`SMART_ALH_CAPTURE_NAME', `SDW2-Capture')
+', `')
```
[the remark this answers is not in the document]
Comment (kv2019i) [reply tools/topology/sof-tgl-sdw-max98373-rt5682.m4:60]: I'm fine with both, but maybe it's more straightforward to keep these separate (now the this is so far ahead already).

**Review by kv2019i — APPROVED** [review]
Good with me.

**Review by plbossart — APPROVED** [review]

**Comment by yongzhi1** [comment]
> @kv2019i @juimonen please review for BT.

For BT offload, this will be added, I can post after this PR get merged:
https://github.com/yongzhi1/sof/commit/87c6c149670ec079a0bb2d76b7ee7dd62619c3c5

**Review by lgirdwood — APPROVED** [review]

**Pull request merged**